### PR TITLE
Bravo 1850 new fees

### PR DESCRIPTION
**Pull request by disenotov**

### CODE REVIEW

File: libraries/contracts/LibOrder.sol [hunk 1]
```diff
@@ -46,7 +47,7 @@ library LibOrder {
 
     function hashKey(Order memory order) internal pure returns (bytes32) {
         //order.data is in hash for V2 orders
-        if (order.dataType == LibOrderDataV2.V2){
+        if (order.dataType == LibOrderDataV2.V2 || order.dataType == LibOrderDataV3.V3_SELL || order.dataType == LibOrderDataV3.V3_BUY){
```
Comment (evgenynacu) [inline libraries/contracts/LibOrder.sol:50]: probably we can revert if here? so we hash with data always except older orders
not sure if that's ok. need to think

File: libraries/contracts/LibDeal.sol [hunk 2]
```diff
@@ -14,4 +15,10 @@ library LibDeal {
         address proxy;
         address from;
     }
+
+    struct DealData {
+        uint protocolFee;
+        uint maxFeesBasePoint;
```
Comment (evgenynacu) [inline libraries/contracts/LibDeal.sol:21]: And this too? Can we lower the size?

File: transfer-manager/contracts/RaribleTransferManager.sol [hunk 3]
```diff
@@ -218,12 +217,17 @@ abstract contract RaribleTransferManager is OwnableUpgradeable, ITransferManager
     function calculateTotalAmount(
```
Comment (evgenynacu) [inline transfer-manager/contracts/RaribleTransferManager.sol:217]: could you add a description for the function?

File: transfer-manager/contracts/RaribleTransferManager.sol [hunk 4]
```diff
@@ -65,17 +65,16 @@ abstract contract RaribleTransferManager is OwnableUpgradeable, ITransferManager
     function doTransfers(
```
Comment (evgenynacu) [inline transfer-manager/contracts/RaribleTransferManager.sol:65]: description

File: exchange-v2/contracts/ExchangeV2Core.sol [hunk 5]
```diff
@@ -109,14 +111,82 @@ abstract contract ExchangeV2Core is Initializable, OwnableUpgradeable, AssetMatc
         emit Match(newFill.rightValue, newFill.leftValue);
     }
 
+    function getMaxFee(
```
Comment (evgenynacu) [inline exchange-v2/contracts/ExchangeV2Core.sol:114]: add description

File: exchange-v2/contracts/ExchangeV2Core.sol [hunk 6]
```diff
@@ -109,14 +111,82 @@ abstract contract ExchangeV2Core is Initializable, OwnableUpgradeable, AssetMatc
...
+            leftDataType,
+            rightDataType,
+            leftOrderData,
+            rightOrderData,
+            dealData.feeSide,
+            dealData.protocolFee
+        );
+    }
+
     function getFillSetNew(
```
Comment (evgenynacu) [inline exchange-v2/contracts/ExchangeV2Core.sol:182]: add description

**Review by evgenynacu — CHANGES_REQUESTED** [review]

File: libraries/contracts/LibOrder.sol (hunk 1 above)
Comment (disenotov) [reply libraries/contracts/LibOrder.sol:50]: reversed the order of if statements

File: libraries/contracts/LibDeal.sol [hunk 7]
```diff
@@ -14,4 +15,10 @@ library LibDeal {
         address proxy;
         address from;
     }
+
+    struct DealData {
+        uint protocolFee;
```
[the remark this answers is not in the document]
Comment (disenotov) [reply libraries/contracts/LibDeal.sol:20]: doesn't save memory

File: libraries/contracts/LibDeal.sol (hunk 2 above)
Comment (disenotov) [reply libraries/contracts/LibDeal.sol:21]: doesn't save memory

File: transfer-manager/contracts/RaribleTransferManager.sol (hunk 3 above)
Comment (disenotov) [reply transfer-manager/contracts/RaribleTransferManager.sol:241]: done

File: transfer-manager/contracts/RaribleTransferManager.sol [hunk 8]
```diff
@@ -86,10 +85,11 @@ abstract contract RaribleTransferManager is OwnableUpgradeable, ITransferManager
     function doTransfersWithFees(
```
[the remark this answers is not in the document]
Comment (disenotov) [reply transfer-manager/contracts/RaribleTransferManager.sol:101]: done

File: exchange-v2/contracts/ExchangeV2Core.sol [hunk 9]
```diff
@@ -163,5 +233,12 @@ abstract contract ExchangeV2Core is Initializable, OwnableUpgradeable, AssetMatc
 
     function getProtocolFee() internal virtual view returns(uint);
 
+    function getProtocolFeeConditional(bytes4 leftDataType) internal view returns(uint) {
```
[the remark this answers is not in the document]
Comment (disenotov) [reply exchange-v2/contracts/ExchangeV2Core.sol:264]: done

File: exchange-v2/contracts/ExchangeV2Core.sol [hunk 10]
```diff
@@ -55,18 +55,14 @@ abstract contract ExchangeV2Core is Initializable, OwnableUpgradeable, AssetMatc
 
     function matchAndTransfer(LibOrder.Order memory orderLeft, LibOrder.Order memory orderRight) internal {
```
[the remark this answers is not in the document]
Comment (disenotov) [reply exchange-v2/contracts/ExchangeV2Core.sol:61]: done

**Pull request merged**